### PR TITLE
[pigeon] Fix C++ generator's handling of Flutter APIs

### PR DESCRIPTION
**Pull request by stuartmorgan**
This is the Flutter API analog to https://github.com/flutter/packages/pull/2270. It overhauls C++ Flutter API generation so that it works for types that aren't classes, as it was originally written when Pigeon didn't support non-data-class arguments. It makes the types look mostly like the host API type mappings introduced in that PR. It also fixes the method names to match the host API names, and C++ style.

It also changes the callback definition for Flutter API calls, as future-proofing for https://github.com/flutter/flutter/issues/118243 (as with other languages, it's not actually used right now since the Dart generator never sends errors). There were several possible options:
1. One callback with an `ErrorOr<T>`. I decided against that since although we use `ErrorOr` for host API return values, clients generally don't have to actually interact with its API; they just return either the right value, or a `FlutterError`. Consuming an `ErrorOr` requires everyone to learn its API, and then write the same conditional every time.
2. One callback with both an optional value and an optional `FlutterError`. This would match the Obj-C generator, but "possibly null error" isn't a standard pattern in C++ the way it is in Obj-C, and the potential confusion with C++ types in how we would pass an optional non-error return value that was nullable would be messy.
3. Two callbacks, one for each. This requires a little more boilerplate in the call, but since it doesn't require any extraction to use the arguments it's simpler in some ways than the boilerplate conditional people would have to write with the options above.

I picked 3, but I'm open to arguments for the other options.

As with https://github.com/flutter/packages/pull/2270 it doesn't specifically attempt to fix enum types; those can be done for both sides in a follow up.

On the testing side:
- Adds generator unit tests similar to the ones done for host API that makes assertions about the specific way types are expressed (since in C++ we have multiple options that would compile, and want to assert that we're using the specific ones as they are chosen for a reason.)
- Adds a lot of new Flutter API definitions to pigeons/core_tests.dart in order to get fairly comprehensive compilation testing for these changes.
- It enables the `callFlutterEchoString` test for Windows now that it works.
- It does *not* add new `callFlutter*` methods and integration tests for all the new Flutter API methods. While we certainly want those tests, I think it's better to do those in a follow-up as this is already a large PR, and adding that would require a ton of new code in every native test.
- Similarly it doesn't add new Windows native unit tests of Flutter API calls. I would prefer to do those after we've revisited how we are structuring our unit tests (i.e., not having them use N different pigeons/ files) to avoid churn there.

Deferring some of the test backfill does mean that there's a potential that some cases are still wrong, but since it's definitely less wrong than it was before, and coverage is improved by this PR even if it's not where we want it yet, this seemed like a good cut point. Adding more tests to this PR is an option though.

This also fixes a minor issue in the Swift generator exposed by one of the new pigeons/ entries.

Fixes https://github.com/flutter/flutter/issues/108682 (modulo caveat above).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
As discussed offline I won't land this until after the main restructuring work has landed to avoid mass conflicts in that PR, but I wanted to get it up for review in parallel with the structural design discussions.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:378]: This code was moved to `_wrappedHostApiArgumentExpression` below for sharing with Flutter API code that needed to do the same thing.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:446]: This used to be the locally-defined `extractEncodedArgument` below, and was similarly extracted for sharing with Flutter API code.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:874]: This code was replaced with `_writeEncodedArgumentExtraction`, which was more comprehensive/correct.

File: packages/pigeon/lib/swift_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:341]: This fixes the warning exposed by the multiple-argument Flutter API method I added to core_tests.dart. The problem is that `[somethingOfTypeA, somethingOfTypeB]` implicitly discards type information about the elements because it becomes an `[Any]` (or `[Any?]`). You have to make it explicit that you know you are doing that to silence the warning.

File: packages/pigeon/platform_tests/shared_test_plugin_code/lib/integration_tests.dart
Comment (stuartmorgan) [inline packages/pigeon/platform_tests/shared_test_plugin_code/lib/integration_tests.dart:596]: These aren't used yet, but had to be implemented since they are now part of the interface.

File: packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.cpp
Comment (stuartmorgan) [inline packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.cpp:548]: This churn is from a minor drive-by streamlining I did: the `*Channel` classes default to `flutter::EncodableValue` as the template type, so `flutter::BasicMessageChannel<>` is equivalent to `flutter::BasicMessageChannel<flutter::EncodableValue>>`, but shorter.

File: packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.cpp
Comment (stuartmorgan) [inline packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.cpp:1514]: This is showing the main fix: this is the old `EchoString` implementation, and it was trying to use `CustomEncodableValue`, which is only for the custom data classes, so won't actually work. Now it's an `EncodableValue` as expected.

File: packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.h
Comment (stuartmorgan) [inline packages/pigeon/platform_tests/test_plugin/windows/pigeon/core_tests.gen.h:379]: This is the change to callbacks to allow for error handling.

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:446]: nit: I'm not sure what "extraction" means based on the docstring.  Is that a term I'm unfamiliar with?

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:761]: This what you are referring to in the description right?  Your choice to have 2 methods here instead of the argument to the callback being ErrorOr, right?  This sounds good to me.  The only downside is that clients would have to factor out shared logic (the equivalent to `finally` blocks).

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:769]: nit: looks like every call to `getFieldHostDatatype` pulls out the classes and enums field.  It would be cleaner probably just pass the AST if that's the case.

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:773]: nit: no need to make a new closure, you should be able to just pass `_baseCppTypeForBuiltinDartType`

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:869]: nit: We can move these callbacks instead of copying them.  Who knows what clients will have in them.  If it is a pain, feel free to leave for later with a note.

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:593]: Is it safe to have a pointer to the reference to the codec in the channel?

File: packages/pigeon/test/cpp_generator_test.dart
Comment (gaaclarke) [inline packages/pigeon/test/cpp_generator_test.dart:1021]: Once a test gets over 100 lines of code, maybe we should have a docstring that explains the test a bit more.  I'm not sure what "map correctly" is immediately without reading through the test.

**Review by gaaclarke — COMMENTED** [review]
Code looks good for the most part modulo a few nits.  Apropos the question about 2 callbacks, that sounds good to me too.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/cpp_generator.dart]: Done. I also changed the way the paramater list was built in both locations to eliminate unnecessary branching for void vs non-void functions.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/cpp_generator.dart]: Fixed.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/cpp_generator.dart]: Done; I tried to do this with `reply` before realizing we need to keep it for fallback error handling, and I just forgot to do it here.

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/cpp_generator.dart]: Yea, thats a hard one to name.  I think unwrapping is more in line with what we call it elsewhere.  The names not going to be clear, we just need to make sure what you just said is conveyed in the docstring.

**Review by gaaclarke — APPROVED** [review]
lgtm

**Comment by stuartmorgan** [comment]
I think I got everything merged correctly into the new structure. The generated output doesn't have any unexpected changes relative to the previous PR snapshot.

**Comment by tarrinneal** [comment]
can you run the formatter and re-upload the updated generated files? Seems you need to run analyze as well, but that might go away with a format.

**Comment by stuartmorgan** [comment]
I missed that the Dart unit tests weren't compiling (the error spam from the C++ and Java projects in VS Code makes it hard to see actual issues in the bottom area, and causes it to always pop up a dialog about their being errors before running unit tests, so it's kind of a pain to tell.) I think that may have been the root cause of all of the CI failures, so I'll see if anything else is still broken.

**Pull request merged**
